content/articles/online-machine-learning-with-river-python/index.md

```diff
@@ -0,0 +1,446 @@
+
+```python
+pipe_nb.steps
+```
+
+Output:
+
+```bash
+OrderedDict([('vectorizer',
+              BagOfWords (
+                on=None
+                strip_accents=True
+                lowercase=True
+                preprocessor=None
+                tokenizer=<built-in method findall of re.Pattern object at 0x7fa35529de00>
+                ngram_range=(1, 1)
+              )),
+             ('nb',
+              MultinomialNB (
+                alpha=1.
+              ))])
+```
+
+### Building our model
+
+Since we are dealing with streaming data, we have to learn from our data one a time to simulate streaming data.
+
+Our data is in form of a list of tuples so we can learn from it one a time by iterating through our data using a for a loop. This will ensure that we train a model to give each instance in our loop over time before going to the next instance as if we are dealing with real-time data that comes in streams.
+
+#### Looping through our dataset
+
+```python
+for text,label in data:
+    pipe_nb = pipe_nb.learn_one(text,label)
+```
+
+When looping through our dataset we use the `learn_one()` method to learn one at a time from our given list of the tuple. `learn_one()` will learn from the first `text` and `label` which is `"my python program is running", "software"` according to our data set.
+
+It will then store this knowledge learned and use it when the next data arrives. Since the model remembers the knowledge gained over time, it will be able to adapt to changes in the dataset. Over time, the model will be more accurate than the beginning of the training since it would have accumulated knowledge.
+
+### Making prediction
+
+Since this is online machine learning, the model does not wait till the end to make predictions. It will predict this instance and continue training when the next data arrives.
+
+```python
+pipe_nb.predict_one("I built an API")
+```
+
+We use `predict_one` to predict according to this instance.
+
+The output is shown:
+
+```bash
+'software'
+```
+
+This is true since our text is `software` related.
+
+### Prediction probability
+
+To get the probability of the above classification use this command.
+
+```python
+pipe_nb.predict_proba_one("I built an API")
+```
+
+We use the `predict_proba_one()` method to get the probability of prediction at this instance.
+The output is as shown.
+
+```python
+{'software': 0.732646964375691, 'hardware': 0.2673530356243093}
+```
+
+This shows that the probability of the text being classified as `software` is more than that of `hardware`. Their probabilities are `0.7326` and `0.2673` respectively.
+
+Online machine learning is a continuous process since we are real-time time data that is continuously being generated.
+
+Let's make another prediction.
+
+#### Other prediction
+
+```python
+pipe_nb.predict_one("the hard drive  in the computer is damaged")
+```
+
+The prediction output:
+
+```bash
+'software'
+```
+
```

Comment:
This is a wrong prediction, right? 

Start by stating that then you can continue with the explanation.